src/lib/Key.js

```diff
@@ -78,14 +78,48 @@ class Key { // eslint-disable-line no-unused-vars
     /**
      * Sign a generic message.
      *
-     * @param {string} message - A utf-8 string
-     * @returns {{message: string, signature: Nimiq.Signature}}
+     * @param {string | Uint8Array} message - A utf-8 string or byte array (max 255 bytes)
+     * @returns {SignedMessageResult}
      */
     signMessage(message) {
-        message = `nimiq_msg_${message}`;
-        const msgBytes = Utf8Tools.stringToUtf8ByteArray(message);
-        const signature = Nimiq.Signature.create(this.keyPair.privateKey, this.keyPair.publicKey, msgBytes);
-        return { message, signature };
+        /**
+         * Adding a prefix to the message makes the calculated signature recognisable as
+         * a Nimiq specific signature. This prevents misuse where a malicious request can
+         * sign arbitrary data (e.g. a transaction) and use the signature to impersonate
+         * the victim. (https://github.com/ethereum/wiki/wiki/JSON-RPC#eth_sign)
+         */
+        const prefix = 'Nimiq Signed Message:\n';
+        const prefixBytes = new Nimiq.SerialBuffer(Utf8Tools.stringToUtf8ByteArray(prefix));
+        const prefixLength = prefixBytes.length;
+
+        /** @type {Nimiq.SerialBuffer} */
+        let msgBytes;
+        if (typeof message === 'string') {
+            msgBytes = new Nimiq.SerialBuffer(Utf8Tools.stringToUtf8ByteArray(message));
+        } else {
+            msgBytes = new Nimiq.SerialBuffer(message);
+        }
+
+        const msgLength = msgBytes.length;
+
+        if (msgLength > 255) {
+            throw new Error('Message must not exceed 255 bytes');
+        }
+
+        const bufLength = 1 /* prefixBytes length value */
+            + prefixLength
+            + 1 /* msgBytes length value */
+            + msgLength;
+
+        // Construct buffer
+        const buf = new Nimiq.SerialBuffer(bufLength);
+        buf.writeUint8(prefixLength);
+        buf.write(prefixBytes);
```

Comment:
There is also SerialBuffer.writeVarLengthStringth

Reply:
Yes, but it only supports ASCII characters and we want to support the full UTF-8 range.

Reply:
Also, I need to know the length of the prefix string anyway to be able to size the buffer (yes, hardcoding would work but that makes it more inflexible).